File: largest-rectangle-in-histogram/largest-rectangle-in-histogram.c
#define MAX(a,b) (((a)>(b))?(a):(b))

int calcArea(int* heights, int begin, int end){
    if (begin > end) {
        return 0;
    }
    int minIndex = begin;
    for (int i = begin + 1; i <= end; i++) {
        if (heights[i] < heights[minIndex]) {
            minIndex = i;
        }
    }
    int max = MAX(calcArea(heights, begin, minIndex - 1),
                 calcArea(heights, minIndex + 1, end));
    return MAX(heights[minIndex] * (end - begin + 1), max);
    
}

int largestRectangleArea(int* heights, int heightsSize){
    return calcArea(heights, 0, heightsSize - 1);
}

File: hard/largest-rectangle-in-histogram/largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function (heights) {
  const stack = [];
  let result = 0;
  heights.push(0);
  for (let i = 0; i < heights.length; i++) {
    while (stack.length && heights[i] < heights[stack[stack.length - 1]]) {
      const height = heights[stack[stack.length - 1]];
      stack.pop();
      const left = stack.length ? stack[stack.length - 1] : -1;
      result = Math.max(result, (i - left - 1) * height);
    }
    stack.push(i);
  }
  return result;
};
